File: lib/core/security/correction.dart
/// Validation for the "report a correction" form.
///
/// This is the Dart twin of `functions/lib/correction.js` (the endpoint,
/// which is authoritative) and `site/lib/correction.ts` (the web form).
/// The three must agree: a change to a bound or an allowed value here
/// needs the identical change in the other two, plus a matching test
/// case in each suite.
///
/// The form runs this before it will submit, so the user sees inline
/// errors; the endpoint runs its own copy again because it cannot trust
/// that the client did.
library;

import 'safe_url.dart';

/// Which part of a profile a report is about.
enum CorrectionField { biography, controversy, sentiment, image, other }

/// The field names as sent to the endpoint. Must match
/// `CORRECTION_FIELDS` in the JavaScript and TypeScript copies.
extension CorrectionFieldWire on CorrectionField {
  String get wire => switch (this) {
    CorrectionField.biography => 'biography',
    CorrectionField.controversy => 'controversy',
    CorrectionField.sentiment => 'sentiment',
    CorrectionField.image => 'image',
    CorrectionField.other => 'other',
  };
}

/// A validated, normalised correction report, ready to serialise.
class CleanCorrection {
  const CleanCorrection({
    required this.slug,
    required this.field,
    required this.claim,
    required this.correction,
    this.evidenceUrl,
    this.email,
  });

  final String slug;
  final CorrectionField field;
  final String claim;
  final String correction;
  final String? evidenceUrl;
  final String? email;

  Map<String, dynamic> toJson() => {
    'slug': slug,
    'field': field.wire,
    'claim': claim,
    'correction': correction,
    if (evidenceUrl != null) 'evidenceUrl': evidenceUrl,
    if (email != null) 'email': email,
  };
}

/// Thrown for the first problem found. [field] names the offending input
/// so the form can highlight it.
class CorrectionError implements Exception {
  const CorrectionError(this.field, this.code, this.message);

  final String field;
  final String code;
  final String message;

  @override
  String toString() => 'CorrectionError($field, $code): $message';
}

/// Bounds, kept identical to the JavaScript and TypeScript copies.
abstract final class CorrectionLimits {
  static const int slugMaxLength = 90;
  static const int claimMinLength = 10;
  static const int claimMaxLength = 600;
  static const int correctionMinLength = 10;
  static const int correctionMaxLength = 1000;
  static const int emailMaxLength = 254;
}

final RegExp _slugPattern = RegExp(r'^[a-z0-9]+(?:-[a-z0-9]+)*$');
final RegExp _emailPattern = RegExp(r'^[^\s@]+@[^\s@]+\.[^\s@]+$');
final RegExp _whitespace = RegExp(r'\s+');

const List<String> _injectionMarkers = [
  'ignore previous',
  'ignore prior',
  'ignore all',
  'disregard',
  'system prompt',
  'you are now',
  'new instructions',
  'act as',
];

String _collapse(String v) => v.replaceAll(_whitespace, ' ').trim();

void _assertNoInjection(String field, String text) {
  final lowered = text.toLowerCase();
  if (_injectionMarkers.any(lowered.contains)) {
    throw CorrectionError(
      field,
      'looks_like_instruction',
      'That reads as an instruction rather than a description.',
    );
  }
}

CorrectionField? _parseField(String wire) {
  for (final f in CorrectionField.values) {
    if (f.wire == wire) return f;
  }
  return null;
}

/// Validates and normalises a correction report. Throws
/// [CorrectionError] on the first problem.
CleanCorrection validateCorrection({
  String? slug,
  String? field,
  String? claim,
  String? correction,
  String? evidenceUrl,
  String? email,
}) {
  final cleanSlug = (slug ?? '').trim().toLowerCase();
  if (cleanSlug.isEmpty) {
    throw const CorrectionError('slug', 'missing', 'No profile was named.');
  }
  if (cleanSlug.length > CorrectionLimits.slugMaxLength ||
      !_slugPattern.hasMatch(cleanSlug)) {
    throw const CorrectionError(
      'slug',
      'invalid',
      'That profile id is not valid.',
    );
  }

  final fieldValue = (field ?? '').trim().toLowerCase();
  final parsedField = _parseField(fieldValue);
  if (parsedField == null) {
    throw const CorrectionError(
      'field',
      'invalid',
      'Choose which part of the profile is wrong.',
    );
  }

  final cleanClaim = _collapse(claim ?? '');
  if (cleanClaim.length < CorrectionLimits.claimMinLength) {
    throw const CorrectionError(
      'claim',
      'too_short',
      'Quote the part you are disputing.',
    );
  }
  if (cleanClaim.length > CorrectionLimits.claimMaxLength) {
    throw CorrectionError(
      'claim',
      'too_long',
      'Keep the disputed text under '
          '${CorrectionLimits.claimMaxLength} characters.',
    );
  }
  _assertNoInjection('claim', cleanClaim);

  final cleanCorrection = _collapse(correction ?? '');
  if (cleanCorrection.length < CorrectionLimits.correctionMinLength) {
    throw const CorrectionError(
      'correction',
      'too_short',
      'Say what it should say instead.',
    );
  }
  if (cleanCorrection.length > CorrectionLimits.correctionMaxLength) {
    throw CorrectionError(
      'correction',
      'too_long',
      'Keep the correction under '
          '${CorrectionLimits.correctionMaxLength} characters.',
    );
  }
  _assertNoInjection('correction', cleanCorrection);

  String? cleanUrl;
  final rawUrl = (evidenceUrl ?? '').trim();
  if (rawUrl.isNotEmpty) {
    final parsed = SafeUrl.parse(rawUrl);
    if (parsed == null) {
      throw const CorrectionError(
        'evidenceUrl',
        'unsafe',
        'The evidence link must be a plain https web address.',
      );
    }
    cleanUrl = parsed.toString();
  }

  String? cleanEmail;
  final rawEmail = (email ?? '').trim();
  if (rawEmail.isNotEmpty) {
    if (rawEmail.length > CorrectionLimits.emailMaxLength ||
        !_emailPattern.hasMatch(rawEmail)) {
      throw const CorrectionError(
        'email',
        'invalid',
        'That email address is not valid.',
      );
    }
    cleanEmail = rawEmail;
  }

  return CleanCorrection(
    slug: cleanSlug,
    field: parsedField,
    claim: cleanClaim,
    correction: cleanCorrection,
    evidenceUrl: cleanUrl,
    email: cleanEmail,
  );
}

File: functions/lib/correction.js
"use strict";

/**
 * Validation for "report a correction" submissions.
 *
 * This is the authoritative copy of the correction-form rules. Two other
 * copies must agree with it: `site/lib/correction.ts` (the web form, which
 * runs the same checks before it will submit) and
 * `lib/core/security/correction.dart` (the Flutter form). A change to a
 * bound or an allowed value here needs the identical change in both, plus
 * a matching test case in each suite.
 *
 * The endpoint cannot trust that a client ran its copy, so it runs this
 * one again on every request.
 */

const {parseSafeUrl} = require("./safeUrl");

const CORRECTION_FIELDS = [
  "biography",
  "controversy",
  "sentiment",
  "image",
  "other",
];

/**
 * What the submission is.
 *
 *   "correction" — anyone reporting that something on the profile is
 *                  wrong. The default, and the historic behaviour.
 *   "response"   — the subject of the profile (or their representative)
 *                  giving their side of a claim that is rendered. Stored
 *                  the same way; a moderator may then attach it to the
 *                  profile so it shows inline beside the claim, which is
 *                  the point of the editorial position: a named living
 *                  person gets to be heard on their own page.
 */
const CORRECTION_KINDS = ["correction", "response"];

const SLUG_MAX_LENGTH = 90;
const CLAIM_MIN_LENGTH = 10;
const CLAIM_MAX_LENGTH = 600;
const CORRECTION_MIN_LENGTH = 10;
const CORRECTION_MAX_LENGTH = 1000;
const EMAIL_MAX_LENGTH = 254;

/** Matches the slugs `validate.toSlug` produces, including `n-<hex>`. */
const SLUG_PATTERN = /^[a-z0-9]+(?:-[a-z0-9]+)*$/;
const EMAIL_PATTERN = /^[^\s@]+@[^\s@]+\.[^\s@]+$/;

/** Free text that reads as an instruction to a model, not as a report. */
const INJECTION_MARKERS = [
  "ignore previous",
  "ignore prior",
  "ignore all",
  "disregard",
  "system prompt",
  "you are now",
  "new instructions",
  "act as",
];

/** A rejected submission. `field` names the offending input. */
class CorrectionError extends Error {
  /** @param {string} field @param {string} code @param {string} message */
  constructor(field, code, message) {
    super(message);
    this.name = "CorrectionError";
    this.field = field;
    this.code = code;
    this.status = 400;
  }
}

/** @param {unknown} v @return {string} */
function asString(v) {
  return typeof v === "string" ? v : "";
}

/** @param {string} v @return {string} */
function collapseWhitespace(v) {
  return v.replace(/\s+/gu, " ").trim();
}

/** @param {string} field @param {string} text */
function assertNoInjection(field, text) {
  const lowered = text.toLowerCase();
  if (INJECTION_MARKERS.some((m) => lowered.includes(m))) {
    throw new CorrectionError(
        field,
        "looks_like_instruction",
        "That reads as an instruction rather than a description.",
    );
  }
}

/**
 * Validates and normalises a correction submission. Throws
 * {@link CorrectionError} on the first problem.
 *
 * @param {Record<string, unknown>} input
 * @return {{slug: string, field: string, claim: string,
 *   correction: string, evidenceUrl: string|null, email: string|null}}
 */
function validateCorrection(input) {
  const raw = input && typeof input === "object" ? input : {};

  const slug = asString(raw.slug).trim().toLowerCase();
  if (slug.length === 0) {
    throw new CorrectionError("slug", "missing", "No profile was named.");
  }
  if (slug.length > SLUG_MAX_LENGTH || !SLUG_PATTERN.test(slug)) {
    throw new CorrectionError(
        "slug", "invalid", "That profile id is not valid.",
    );
  }

  const field = asString(raw.field).trim().toLowerCase();
  if (!CORRECTION_FIELDS.includes(field)) {
    throw new CorrectionError(
        "field", "invalid", "Choose which part of the profile is wrong.",
    );
  }

  // Optional; absent means a third-party correction, the original
  // behaviour.
  const kindRaw = asString(raw.kind).trim().toLowerCase();
  const kind = kindRaw === "" ? "correction" : kindRaw;
  if (!CORRECTION_KINDS.includes(kind)) {
    throw new CorrectionError("kind", "invalid", "Unknown submission type.");
  }

  const claim = collapseWhitespace(asString(raw.claim));
  if (claim.length < CLAIM_MIN_LENGTH) {
    throw new CorrectionError(
        "claim", "too_short", "Quote the part you are disputing.",
    );
  }
  if (claim.length > CLAIM_MAX_LENGTH) {
    throw new CorrectionError(
        "claim",
        "too_long",
        `Keep the disputed text under ${CLAIM_MAX_LENGTH} characters.`,
    );
  }
  assertNoInjection("claim", claim);

  const correction = collapseWhitespace(asString(raw.correction));
  if (correction.length < CORRECTION_MIN_LENGTH) {
    throw new CorrectionError(
        "correction", "too_short", "Say what it should say instead.",
    );
  }
  if (correction.length > CORRECTION_MAX_LENGTH) {
    throw new CorrectionError(
        "correction",
        "too_long",
        `Keep the correction under ${CORRECTION_MAX_LENGTH} characters.`,
    );
  }
  assertNoInjection("correction", correction);

  let evidenceUrl = null;
  const rawUrl = asString(raw.evidenceUrl).trim();
  if (rawUrl.length > 0) {
    const parsed = parseSafeUrl(rawUrl);
    if (parsed === null) {
      throw new CorrectionError(
          "evidenceUrl",
          "unsafe",
          "The evidence link must be a plain https web address.",
      );
    }
    evidenceUrl = parsed.toString();
  }

  let email = null;
  const rawEmail = asString(raw.email).trim();
  if (rawEmail.length > 0) {
    if (rawEmail.length > EMAIL_MAX_LENGTH || !EMAIL_PATTERN.test(rawEmail)) {
      throw new CorrectionError(
          "email", "invalid", "That email address is not valid.",
      );
    }
    email = rawEmail;
  }

  return {slug, field, kind, claim, correction, evidenceUrl, email};
}

module.exports = {
  validateCorrection,
  CorrectionError,
  CORRECTION_FIELDS,
  CORRECTION_KINDS,
  SLUG_MAX_LENGTH,
  CLAIM_MIN_LENGTH,
  CLAIM_MAX_LENGTH,
  CORRECTION_MIN_LENGTH,
  CORRECTION_MAX_LENGTH,
  EMAIL_MAX_LENGTH,
};
